File: sdks/node/src/plugins/qwen-installation.ts
import path from 'node:path';
import type { InstallConfig } from './base.js';
import {
  AGENT_KEY,
  AUDIT_SCRIPT,
  GUARD_SCRIPT,
} from './qwen-contract.js';
import { sameQwenPath } from './qwen-command.js';
import {
  qwenDocumentLabel,
  type RenderedQwenDocument,
} from './qwen-config.js';
import {
  commitManagedInstallation,
  preflightManagedInstallation,
  prepareManagedInstallation,
  type ManagedRuntimePaths,
  type PreparedManagedInstallation,
  type RenameFile,
} from './managed-installation.js';
import {
  commitManagedTransaction,
  prepareManagedFileChange,
  type PreparedManagedTransaction,
} from './managed-transaction.js';
import {
  requireQwenHooksEnabled,
  type QwenSources,
} from './qwen-sources.js';

const DISPLAY_NAME = 'Qwen Code';
const SETTINGS_DIRECTORY_LABEL = 'Qwen Code user configuration directory';
const MAX_SOURCE_BYTES = 2 * 1024 * 1024;

export type QwenRuntimePaths = ManagedRuntimePaths;
export type PreparedQwenInstallation = PreparedManagedInstallation;
export type { RenameFile };

export async function preflightQwenInstallation(
  config: InstallConfig,
  sources: QwenSources,
): Promise<QwenRuntimePaths> {
  requireQwenHooksEnabled(sources);
  return preflightManagedInstallation({
    agentKey: AGENT_KEY,
    hookLocations: [{
      directoryLabel: SETTINGS_DIRECTORY_LABEL,
      filePath: sources.user.filePath,
    }],
    config,
  }, GUARD_SCRIPT, AUDIT_SCRIPT);
}

function readOnlyPreconditions(sources: QwenSources, changedPath?: string) {
  return sources.preconditions.filter((condition) => (
    changedPath === undefined || !sameQwenPath(condition.filePath, changedPath)
  ));
}

export async function prepareQwenInstallation(
  config: InstallConfig,
  sources: QwenSources,
  rendered: RenderedQwenDocument,
): Promise<PreparedQwenInstallation> {
  await preflightQwenInstallation(config, sources);
  const settingsSource = rendered.next ?? rendered.document.raw;
  const prepared = await prepareManagedInstallation({
    agentKey: AGENT_KEY,
    displayName: DISPLAY_NAME,
    hookSources: [{
      directoryLabel: SETTINGS_DIRECTORY_LABEL,
      label: qwenDocumentLabel(rendered.document),
      filePath: rendered.document.filePath,
      expectedSource: rendered.document.exists ? rendered.document.raw : undefined,
      expectedSnapshot: rendered.document.snapshot,
      source: settingsSource,
    }],
    config,
    auditOptions: { nativePayload: true },
  }, GUARD_SCRIPT, AUDIT_SCRIPT);
  return {
    ...prepared,
    transaction: {
      ...prepared.transaction,
      preconditions: readOnlyPreconditions(sources, rendered.document.filePath),
    },
  };
}

export async function commitQwenInstallation(
  prepared: PreparedQwenInstallation,
  renameFile?: RenameFile,
): Promise<void> {
  await commitManagedInstallation(prepared, renameFile);
}

export async function prepareQwenUninstall(
  sources: QwenSources,
  rendered: RenderedQwenDocument,
): Promise<PreparedManagedTransaction> {
  const change = rendered.changed
    ? await prepareManagedFileChange({
      filePath: rendered.document.filePath,
      label: qwenDocumentLabel(rendered.document),
      next: rendered.next,
      mode: 0o600,
      maximumBytes: MAX_SOURCE_BYTES,
      expectedSource: rendered.document.exists ? rendered.document.raw : undefined,
      expectedSnapshot: rendered.document.snapshot,
      verifyExpectedSource: true,
    })
    : undefined;
  return {
    displayName: DISPLAY_NAME,
    operation: 'uninstall',
    directories: [{
      path: path.dirname(rendered.document.filePath),
      label: SETTINGS_DIRECTORY_LABEL,
    }],
    changes: change ? [change] : [],
    preconditions: readOnlyPreconditions(
      sources,
      rendered.changed ? rendered.document.filePath : undefined,
    ),
  };
}

export async function commitQwenUninstall(
  prepared: PreparedManagedTransaction,
  renameFile?: RenameFile,
): Promise<void> {
  await commitManagedTransaction(prepared, renameFile);
}

File: sdks/go/cmd/elydora/plugins/qwen_installation.go
package plugins

import "fmt"

type preparedQwenInstallation struct {
	changes       []*fileChange
	preconditions []filePrecondition
	paths         *managedRuntimePaths
}

func preflightQwenInstallation(
	config InstallConfig,
	sources *qwenSources,
) (*managedRuntimePaths, string, error) {
	if sources == nil || sources.user == nil || sources.user.filePath == "" {
		return nil, "", fmt.Errorf("Qwen Code installation requires user settings")
	}
	if err := requireQwenHooksEnabled(sources); err != nil {
		return nil, "", err
	}
	if err := validateManagedInstallConfig(config, qwenAgentKey, "Qwen Code"); err != nil {
		return nil, "", err
	}
	paths, err := resolveManagedRuntimePaths(config, qwenGuardScript, qwenAuditScript)
	if err != nil {
		return nil, "", err
	}
	if err := validateManagedRuntimeIdentity(
		paths.agentDirectory, config.AgentID, qwenAgentKey, "Qwen Code",
	); err != nil {
		return nil, "", err
	}
	nodePath, err := resolveAbsoluteNodeRuntime("Qwen Code")
	if err != nil {
		return nil, "", err
	}
	return paths, nodePath, nil
}

func qwenExpectedScripts(agentID string, _ map[string]any) ([]byte, []byte) {
	return []byte(generateGuardScript(qwenAgentKey, agentID, "", false, "")),
		[]byte(buildHookScriptWithOutput(qwenAgentKey, agentID, "", false, true))
}

func prepareRenderedQwenChange(
	rendered *qwenRenderedDocument,
) (*fileChange, error) {
	if rendered == nil || !rendered.changed {
		return nil, nil
	}
	return prepareSnapshotSourceChange(
		rendered.document.filePath,
		qwenDocumentLabel(rendered.document),
		rendered.document.snapshot,
		rendered.next,
		0600,
		rendered.remove,
	)
}

func qwenReadOnlyPreconditions(
	sources *qwenSources,
	changedPath string,
) []filePrecondition {
	if sources == nil {
		return nil
	}
	result := make([]filePrecondition, 0, len(sources.preconditions))
	for _, condition := range sources.preconditions {
		if changedPath != "" && sameManagedPath(condition.filePath, changedPath) {
			continue
		}
		result = append(result, condition)
	}
	return result
}

func prepareQwenInstallation(
	config InstallConfig,
	sources *qwenSources,
	rendered *qwenRenderedDocument,
) (*preparedQwenInstallation, error) {
	paths, _, err := preflightQwenInstallation(config, sources)
	if err != nil {
		return nil, err
	}
	if rendered == nil || rendered.document == nil ||
		!sameManagedPath(rendered.document.filePath, sources.user.filePath) {
		return nil, fmt.Errorf("Qwen Code rendered user settings are required")
	}
	guardScript, auditScript := qwenExpectedScripts(config.AgentID, nil)
	changes, err := managedRuntimeFileChanges(
		config, paths, qwenAgentKey, string(guardScript), string(auditScript),
	)
	if err != nil {
		return nil, err
	}
	settingsChange, err := prepareRenderedQwenChange(rendered)
	if err != nil {
		return nil, err
	}
	changes = append(changes, settingsChange)
	changedPath := ""
	if rendered.changed {
		changedPath = rendered.document.filePath
	}
	return &preparedQwenInstallation{
		changes:       changes,
		paths:         paths,
		preconditions: qwenReadOnlyPreconditions(sources, changedPath),
	}, nil
}

func commitQwenInstallation(
	prepared *preparedQwenInstallation,
	rename renameFunc,
) error {
	if prepared == nil || prepared.paths == nil {
		return fmt.Errorf("prepared Qwen Code installation is required")
	}
	if err := EnsurePrivateDirectory(prepared.paths.runtimeRoot); err != nil {
		return err
	}
	if err := EnsurePrivateDirectory(prepared.paths.agentDirectory); err != nil {
		return err
	}
	return writeChanges(
		prepared.changes,
		"Install Qwen Code hooks",
		rename,
		prepared.preconditions...,
	)
}

func prepareQwenUninstall(
	sources *qwenSources,
	rendered *qwenRenderedDocument,
) (*fileChange, []filePrecondition, error) {
	change, err := prepareRenderedQwenChange(rendered)
	if err != nil {
		return nil, nil, err
	}
	changedPath := ""
	if rendered != nil && rendered.changed {
		changedPath = rendered.document.filePath
	}
	return change, qwenReadOnlyPreconditions(sources, changedPath), nil
}

func commitQwenUninstall(
	change *fileChange,
	preconditions []filePrecondition,
	rename renameFunc,
) error {
	return writeChanges(
		[]*fileChange{change},
		"Uninstall Qwen Code hooks",
		rename,
		preconditions...,
	)
}
